presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNormalizeFunction.java

```diff
@@ -0,0 +1,127 @@
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static java.lang.String.format;
+
+@ScalarFunction("array_normalize")
+@Description("Normalizes an array by dividing each element by the p-norm of the array. " +
+        "This is equivalent to TRANSFORM(array, v -> v / REDUCE(array, 0, (a, v) -> a + POW(ABS(v), p), a -> POW(a, 1 / p)), " +
+        "but the reduce part is only executed once. " +
+        "Returns null if the array is null or there are null array elements.")
+public final class ArrayNormalizeFunction
+{
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = double.class)
+    @SqlType("array(T)")
+    @SqlNullable
+    public static Block normalizeDoubleArray(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block block,
+            @SqlType("T") double p)
+    {
+        return normalizeArray(elementType, block, p, ArrayNormalizeFunction::getDoubleValue, ArrayNormalizeFunction::writeDoubleValue);
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = long.class)
+    @SqlType("array(T)")
+    @SqlNullable
+    public static Block normalizeRealArray(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block block,
+            @SqlType("T") long p)
+    {
+        return normalizeArray(elementType, block, Float.intBitsToFloat((int) p), ArrayNormalizeFunction::getRealValue, ArrayNormalizeFunction::writeRealValue);
+    }
+
+    @FunctionalInterface
+    private interface ValueGetter
+    {
+        double getValue(Type elementType, Block block, int i);
+    }
+
+    @FunctionalInterface
+    private interface ValueWriter
+    {
+        void writeValue(Type elementType, BlockBuilder blockBuilder, double value);
+    }
+
+    private static Block normalizeArray(Type elementType, Block block, double p, ValueGetter valueGetter, ValueWriter valueWriter)
+    {
+        if (!(elementType instanceof RealType) && !(elementType instanceof DoubleType)) {
+            throw new PrestoException(FUNCTION_IMPLEMENTATION_MISSING,
+                    format("Unsupported array element type for array_normalize function: %s", elementType.getDisplayName()));
+        }
+        checkCondition(p >= 0, INVALID_FUNCTION_ARGUMENT, "array_normalize only supports non-negative p: %s", p);
+        if (p == 0) {
+            return block;
+        }
+        final int numElements = block.getPositionCount();
+        double pNorm = 0;
+        for (int i = 0; i < numElements; ++i) {
+            if (block.isNull(i)) {
+                return null;
+            }
+            pNorm += Math.pow(Math.abs(valueGetter.getValue(elementType, block, i)), p);
+        }
+        if (pNorm == 0) {
+            return block;
```

Comment:
Can we also cover this codepath in the unit test?

Reply:
This is already covered in the test0Norm test.

Reply:
Added new tests to cover this.